db/index.js

```diff
@@ -22,41 +26,47 @@ const useJson = (() => {
 var DB = (() => {
 
   const validateCode = async (code) => {
-    code = code.toUpperCase()
 
     if (useJson) {
-      return await jsonDB.find(user => user.code === code) || null
+      return await jsonDB.find(user => verifyHash(code.toUpperCase(), user.code, true)) || null
     }
 
+    code = hashString(code, true).catch(e => console.warn(e))
+    
     const { rows } = await pool.query('SELECT * FROM public.access_codes WHERE code = $1', [code])
     
     return rows[0] || null
   }
 
   const validateUser = async ({ code, sin, dateOfBirth }) => {
-    code = code.toUpperCase()
     sin = cleanSIN(sin)
 
     let row
 
     if (useJson) {
-      row = jsonDB.find(user => {
-        if (user.sin === sin && user.dateOfBirth === dateOfBirth) {
-          return user
-        }
-      })
+      code = (code.length === 9) ? hashString(code.toUpperCase(), true) : code
+
+      row = jsonDB.find(user => user.code === code)
     } else {
-      const { rows } = await pool.query('SELECT * FROM public.access_codes WHERE sin = $1 AND dob = $2', [sin, dateOfBirth])
+      //find by access code, and then check sin and dob
+      //this is to save us from needing to go through each entry and verify each hash. Pull by the code we already have, and then check the sin and dob hash
+      //it saves having to use a static salt everywhere (slight security risk), and lets us use a randomly generated salt
+      const { rows } = await pool.query('SELECT * FROM public.access_codes WHERE code = $1', [code])
 
       row = rows[0]
     }
 
-    if (row && row.code !== code) {
-      // code from the returned user doesn't match
-      return {"error": true}
+    if (!row) { return { "error" :  true } }
+
+    const incorrectInfo = [verifyHash(sin, row.sin), verifyHash(dateOfBirth, row.dob)].filter(v => v === false) 
+
+    if(incorrectInfo.length > 1) {
+      return { "error" :  true }
+    } else if (incorrectInfo.length) {
+      return null
     }
```

Comment:
yeah, I tried following along when we send each use to what page and it became a bit confusing to map it all out. I think it was also easier to write before because we found a row by a matching SIn and DoB, but since that would require a ton of looping over each entry and verifying now, I pull in by the existing code we already have (so we only have to look up one row), and then check the SIN + DoB. I'd compare it against what we have now to sort of get what I'm saying 😅 
Happy to change this though.

Reply:
Yeah, let's do another card for this but park it in the context of this PR. All good with this logic for the time being.